bills.py

```diff
@@ -0,0 +1,104 @@
+import csv
+import requests
+from dataclasses import dataclass, field
+from typing import List
+from bs4 import BeautifulSoup
+
+ROOT_DOMAIN: str = "https://www.ola.org/"
+CSV_FILENAME: str = "bills.csv"
+
+
+@dataclass
+class SponsorInfo:
+    name: str
+    title: str = ""  # Default to blank if not available
+
+
+@dataclass
+class BillInfo:
+    parliament: str
+    bill_number: str
+    bill_title: str
+    url_title: str
+    sponsors: List[SponsorInfo] = field(default_factory=list)
+
+
+def extract(parliament_url) -> str:
+    response = requests.get(parliament_url)
+    response.raise_for_status()
+    return response.text
+
+
+def transform(parliament: str, html_content: str) -> List[BillInfo]:
+    soup = BeautifulSoup(html_content, "html.parser")
+    table = soup.find("table", class_="views-table views-view-table cols-3")
+
+    rows = table.find("tbody").find_all("tr")
+    bill_info_list = []
+
+    for row in rows:
+        cols = row.find_all("td")
+        bill_number = cols[0].text.strip()
+        bill_link_element = cols[1].find("a")
+        bill_title = bill_link_element.text.strip() if bill_link_element else ""
+        url_title = (
+            ROOT_DOMAIN + bill_link_element["href"].lstrip("/")
+            if bill_link_element
+            else ""
+        )
+
+        bill_sponsors = []
+        sponsors = cols[2].find_all("article", class_="node--type-bill-sponsor")
+        for sponsor in sponsors:
+            sponsor_name = sponsor.find(
+                "div", class_="field--name-field-full-name-by-last-name"
+            ).text.strip()
+            sponsor_title_element = sponsor.find(
+                "div", class_="field--name-field-sponsor-role"
+            )
+            sponsor_title = (
+                sponsor_title_element.text.strip() if sponsor_title_element else ""
+            )
+            bill_sponsors.append(SponsorInfo(name=sponsor_name, title=sponsor_title))
+
+        bill_info_list.append(
+            BillInfo(parliament, bill_number, bill_title, url_title, bill_sponsors)
+        )
+
+    return bill_info_list
+
+
+def load(bills_data: List[BillInfo]):
+    with open(CSV_FILENAME, "w", newline="", encoding="utf-8") as csvfile:
+        fieldnames = ["parliament", "bill_number", "bill_title", "url_title"]
+        max_sponsors = len(max(bills_data, key=lambda x: len(x.sponsors)).sponsors)
+        for i in range(max_sponsors):
+            fieldnames.extend([f"sponsor_name{i+1}", f"sponsor_title{i+1}"])
+
+        writer = csv.DictWriter(csvfile, fieldnames=fieldnames)
+        writer.writeheader()
+
+        for bill in bills_data:
+            row = {
+                "parliament": bill.parliament,
+                "bill_number": bill.bill_number,
+                "bill_title": bill.bill_title,
+                "url_title": bill.url_title,
```

Comment:
Is there a nicer way of creating a dict for {attribute_name: attribute_value}? it looks fine, I'm just curious if there was an easier way for something that has a lot more fields. It'd also probably be hard to do that with the sponsors list here

Reply:
Yeah, I don't think there's a cleaner way to do it. Usually, I'd be down for some Lambda madness, but I think this much more readable